beacon-chain/operations/blstoexec/pool.go

```diff
@@ -0,0 +1,112 @@
+package blstoexec
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v3/beacon-chain/state"
+	fieldparams "github.com/prysmaticlabs/prysm/v3/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+// PoolManager maintains pending and seen BLS-to-execution-change objects.
+// This pool is used by proposers to insert BLS-to-execution-change objects into new blocks.
+type PoolManager interface {
+	PendingBLSToExecChanges(noLimit bool) []*ethpb.SignedBLSToExecutionChange
+	InsertBLSToExecChange(state state.ReadOnlyBeaconState, change *ethpb.SignedBLSToExecutionChange)
+	MarkIncluded(change *ethpb.SignedBLSToExecutionChange)
+}
+
+// Pool is a concrete implementation of PoolManager.
+type Pool struct {
+	lock    sync.RWMutex
+	pending []*ethpb.SignedBLSToExecutionChange
+}
+
+// NewPool returns an initialized pool.
+func NewPool() *Pool {
+	return &Pool{
+		pending: make([]*ethpb.SignedBLSToExecutionChange, 0),
+	}
+}
+
+// PendingBLSToExecChanges returns objects that are ready for inclusion at the given slot.
+// Without returnAll, this method will not return more than the block enforced MaxBlsToExecutionChanges.
+func (p *Pool) PendingBLSToExecChanges(returnAll bool) []*ethpb.SignedBLSToExecutionChange {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if returnAll {
+		return p.pending
+	}
+	maxChanges := params.BeaconConfig().MaxBlsToExecutionChanges
+	pending := make([]*ethpb.SignedBLSToExecutionChange, maxChanges)
+	for i, ch := range p.pending {
+		pending[i] = ch
+		if uint64(len(pending)) == maxChanges {
+			break
+		}
+	}
+	return pending
+}
+
+// InsertBLSToExecChange inserts an object into the pool.
+// This method is a no-op if the pending exit already exists or the validator is already exited.
+func (p *Pool) InsertBLSToExecChange(state state.ReadOnlyBeaconState, change *ethpb.SignedBLSToExecutionChange) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Prevent malformed messages from being inserted.
+	if change == nil ||
+		change.Message == nil ||
+		len(change.Signature) != fieldparams.BLSSignatureLength ||
+		len(change.Message.FromBlsPubkey) != fieldparams.BLSPubkeyLength ||
+		len(change.Message.ToExecutionAddress) != fieldparams.ExecutionAddressLength {
+		return
+	}
```

Comment:
Why are we having these checks here? the Beacon Api prevents that we even gossip them if we haven't checked the conditions in `process_bls_to_execution_changes`.